### PR TITLE
esp_tinyusb: String descriptor handling

### PR DESCRIPTION
**Pull request by tore-espressif**
There are following changes:

### Commit 816c3ad7d036e6563953cce840ef5c8e039ddd61
* Cleanup TinyUSB string descriptors handling. Remove unused code. Add length check to arguments passed to `tinyusb_driver_install`
* Add private option to modify string descriptors

### Commit  770d9a9204e7a7c5936de5c61f99b2845cc46f89
* Use latest [tinyusb](https://components.espressif.com/components/espressif/tinyusb) v0.14.2
* Update Kconfig and tusb_config.h for new feature (Bluetooth, NET, DFU)

~### Commit 97a108e686134149f250366e0fdd3703e7c80afc~
~* Proposal of MAC string handling for NET class~
~* Will be dropped~

Related #133 

I can split the PR into 2, if it is easier to review.

### CODE REVIEW

**Comment by tore-espressif** [comment]
@roma-jam @esp-saurabhbansal @Jacques-zhao PTAL

File: usb/esp_tinyusb/descriptors_control.c
Comment (roma-jam) [inline usb/esp_tinyusb/descriptors_control.c:137]: According to this code we can add a special string descriptors within the following numbers = [0 .. 7]. 
Based on the theory that any string descriptor can be present in the device configuration (for example 0xEE for Microsoft OS Descriptors support) probably it is better to make the list of the string descriptors with the number, length and string data. 

But if the NCM does not need special string number and mac addr can be saved in slots under 5,6,7 numbers...that can work then.

**Review by roma-jam — COMMENTED** [review]
Hi @tore-espressif, 
Great job! 

I had tried to verify as much as I can and have left a couple of comments. 

There are some thoughts (which probably can be easily treated "as is") but otherwise LGTM.

File: usb/esp_tinyusb/descriptors_control.c
Comment (tore-espressif) [reply usb/esp_tinyusb/descriptors_control.c:132]: Good catch! will be fixed

File: usb/esp_tinyusb/descriptors_control.c
Comment (tore-espressif) [reply usb/esp_tinyusb/descriptors_control.c:137]: Good point, I didn't take these special string descriptors into account.

I haven't encountered too many uses of Microsoft Descriptors, so I'm tempted to postpone this breaking feature...

Do you know some example usages?

File: usb/esp_tinyusb/include_private/descriptors_control.h
Comment (tore-espressif) [reply usb/esp_tinyusb/include_private/descriptors_control.h:28]: Correct, ATM, all descriptors passed to TinyUSB must have full application lifetime. I'll emphasize it in docs

File: usb/esp_tinyusb/tinyusb.c
Comment (tore-espressif) [reply usb/esp_tinyusb/tinyusb.c:77]: This is deliberate. In case someone changes value of `USB_STRING_DESCRIPTOR_ARRAY_SIZE` in the future (to support more string descriptors) or removes it altogether (to support dynamic memory allocation for string descriptors), the backward compatibility with `esp_tinyusb v1.0.0` will not be broken.

Do you agree?

**Review by tore-espressif — COMMENTED** [review]
@roma-jam Thank you for the review!

I left 2 threads open for discussion, the rest will be fixed

File: usb/esp_tinyusb/include_private/descriptors_control.h
Comment (roma-jam) [reply usb/esp_tinyusb/include_private/descriptors_control.h:28]: Honestly, no.
So it is nice to have option to set 0xEE string descriptors but I have never seen them in real life.
But not neccesary I suppose.

File: usb/esp_tinyusb/descriptors_control.c
Comment (roma-jam) [reply usb/esp_tinyusb/descriptors_control.c:137]: Honestly, no.
So it is nice to have option to set 0xEE string descriptors but I have never seen them in real life.
But not neccesary I suppose.

File: usb/esp_tinyusb/tinyusb.c
Comment (roma-jam) [reply usb/esp_tinyusb/tinyusb.c:77]: Yes, sure.

File: usb/esp_tinyusb/descriptors_control.c
Comment (tore-espressif) [reply usb/esp_tinyusb/descriptors_control.c:138]: OK, let's leave it to another PR :)

**Comment by esp-saurabhbansal** [comment]
@tore-espressif 
I have left one comment, please check.

**Review by roma-jam — APPROVED** [review]
LGTM

File: usb/esp_tinyusb/descriptors_control.c
Comment (tore-espressif) [reply usb/esp_tinyusb/descriptors_control.c:133]: Correct, this is one of the main reason for this PR. The underflow can happen with the previous implementation.

To keep backward compatibility, the user can still 'misconfigure' the `esp_tinyusb` component. We should fix the HID example after this PR is merged.

Other option is to introduce a breaking change in functionality.

Comment (esp-saurabhbansal) [reply usb/esp_tinyusb/descriptors_control.c:133]: agree

**Review by esp-saurabhbansal — APPROVED** [review]
LGTM

File: usb/esp_tinyusb/Kconfig
Comment (Dazza0) [inline usb/esp_tinyusb/Kconfig:33]: Nitpick: `comment "You can provide your custom descriptors via tinyusb_driver_install()"`

File: usb/esp_tinyusb/Kconfig
Comment (Dazza0) [inline usb/esp_tinyusb/Kconfig:170]: Nitpick: `Select which DFU driver you want to use.`

**Review by Dazza0 — APPROVED** [review]
@tore-espressif Left a couple of nitpicks.

One more nitpick. Is the [WIP: Draft of MAC string handling](https://github.com/espressif/idf-extra-components/pull/134/commits/be445da17c30a9460057a7f3e309f206f7f3ca65) commit still WIP? Do we need to update the commit message's name.

LGTM otherwise

**Comment by tore-espressif** [comment]
> @tore-espressif Left a couple of nitpicks.
> 
> One more nitpick. Is the [WIP: Draft of MAC string handling](https://github.com/espressif/idf-extra-components/pull/134/commits/be445da17c30a9460057a7f3e309f206f7f3ca65) commit still WIP? Do we need to update the commit message's name.
> 
> LGTM otherwise

Thank you for the review @Dazza0 !

I plan to drop that commit before merging. One of the reasons for the PR is to prepare `esp_tinyusb` for new network drivers. I added that^ commit to show the reviewers how I imagined the upper layer MAC string hadnling

**Pull request merged**
